### PR TITLE
introduce custom filenames for idp/sp cert/key

### LINKED ISSUE #392: settings parser throws "idp cert not found" exception
Hi everyone,

I just found a potential issue in the configuration check that causes a `OneLogin_Saml2_Error` Exception with the message `idp_cert_or_fingerprint_not_found_and_required`.
As with many other programs I wan't to keep my certificate files in a special folder on my server. Therefor I left the x509 cert and key fields in the SP and IDP config part empty and provided a `custom_base_path`. The SP cert and key are loaded correctly from the file, just the configuration check for the IDP config is failing.

**How to reproduce:**
1. set a `custom_base_path` in the config
2. place all files (`sp.crt`, `sp.key`, `idp.crt`) in a `certs` folder in the base path
3. leave all x509 cert and key fileds in the SP and IDP config empty
4. set `signMetadata` to True
5. try to get the Metadata from the SP.

**Suggested Fix**
I identified this line of code as the root cause of the issue, as it only checks for an x509 cert in the IDP config, but not in the file.
https://github.com/SAML-Toolkits/python3-saml/blob/a1211a8695c855b74591a607cf589682307572a6/src/onelogin/saml2/settings.py#L392
My Suggestion would be to use the native function in order to check both the configuration and a potential cert file.
`exists_x509 = bool(self.get_idp_cert())`

If this is considered as an issue that needs to be fixed I'll be happy to open a pull request.

### PR DESCRIPTION
**Pull request by do9xe**
This PR introduces the possibility to have custom filenames for the service provider certificate and key, as well for the IdP cert.

**Background**
When deploying code in production environments it is common to have a dedicated folder for certificates. Especially if a server is hosting several services this folder also can contain several certificates. Having a custom filename enhances usability for administrators and gives the user more flexibility.

**about this PR**
- I introduced three new methods to the settings object class to set the custom filenames
- I had to change the order of the code in `_load_settings_dict()`, because it doesn't matter if we check the config before or after copying it to the protected attributes. Also, this became necessary as the protected attributes are called during config check.
- I added the default values for the filenames
- This closes #392 
- I added some tests in order to test against a custom filename and a non-existing filename

**Tests**
I ran all tests and all relevant tests passed, except for two of them, but reason for this is Windows using a different path separator and I didn't want to change the tests also.

### CODE REVIEW

File: src/onelogin/saml2/settings.py [hunk 1]
```diff
@@ -214,26 +232,27 @@ def get_schemas_path(self):
...
-        Loads settings info from a settings Dict
+        Loads settings info from a settings Dict, adds default values and validates the settings
 
         :param settings: SAML Toolkit Settings
         :type settings: dict
 
         :returns: True if the settings info is valid
         :rtype: boolean
         """
+        self._sp = settings.get('sp', {})
```
Comment (pitbulk) [inline src/onelogin/saml2/settings.py:243]: Order should not be changed.

File: src/onelogin/saml2/settings.py [hunk 2]
```diff
@@ -328,6 +347,11 @@ def _add_default_values(self):
         self._sp.setdefault('x509cert', '')
         self._sp.setdefault('privateKey', '')
 
+        # Set the default filenames for the certificates and keys
+        self._idp.setdefault('cert_filename', 'idp.crt')
```
Comment (pitbulk) [inline src/onelogin/saml2/settings.py:351]: Better inject them in the _add_default_values method

Comment (do9xe) [reply src/onelogin/saml2/settings.py:351]: But it is already part of the `_add_default_values` method

File: src/onelogin/saml2/settings.py (hunk 1 above)
Comment (do9xe) [reply src/onelogin/saml2/settings.py:243]: The order doesn't matter to the functionality, as the '_load_settings_from_dict' method is only called during initialization. To make the code easier I use the `get_idp_cert()` method to check if the config option or file actually exists and has content. Using the old order we would require way more code as the `get_idp_cert()` uses the data from the `_idp` dict (and so does the `get_sp_cert()` method)

**Comment by do9xe** [comment]
Any further comments or can we merge this?

**Comment by do9xe** [comment]
What's the status here? Can this be merged or does this patch need some additional work?

**Review by pitbulk — APPROVED** [review]

**Comment by do9xe** [comment]
> Error running coveralls: Running on Github Actions but GITHUB_TOKEN is not set. Add "env: GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}" to your step config.

Looks like the CI pipeline has a problem

**Comment by do9xe** [comment]
@pitbulk can you please check the CI or merge it manually?

**Pull request merged**